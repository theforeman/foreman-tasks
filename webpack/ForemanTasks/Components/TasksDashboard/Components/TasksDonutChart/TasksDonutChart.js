import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Chart from '../../../../Components/Chart/Chart';

import {
  TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS,
  TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS_ARRAY,
  COLLOR_PATTERN,
} from './TasksDonutChartConstants';
import {
  baseChartConfig,
  shouleBeSelected,
  createChartData,
  updateChartTitle,
} from './TasksDonutChartHelper';
import './TasksDonutChart.scss';

class TasksDonutChart extends React.Component {
  componentDidUpdate() {
    this.updateChartTitle();
    this.updateChartFocus();
  }

  onChartCreate(chart) {
    this.chart = chart;

    this.updateChartTitle();
    this.updateChartFocus();
  }

  updateChartTitle() {
    const { last, older, onTotalClick } = this.props;

    updateChartTitle({
      chartElement: this.chart.element,
      value: last + older,
      onClick: () => onTotalClick(),
      onMouseOver: () =>
        this.setChartFocus(TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.TOTAL),
      onMouseOut: () => this.updateChartFocus(),
    });
  }

  updateChartFocus() {
    this.setChartFocus(this.props.focusedOn);
  }

  setChartFocus(focusedOn) {
    switch (focusedOn) {
      case TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.TOTAL:
        this.chart.focus([
          TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.LAST,
          TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.OLDER,
        ]);
        break;
      case TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.LAST:
      case TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.OLDER:
        this.chart.focus(focusedOn);
        break;
      case TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.NONE:
        this.chart.focus([]);
        break;
      default:
        this.chart.revert();
    }
  }

  render() {
    const {
      className,
      last,
      older,
      timePeriod,
      focusedOn,
      colorsPattern,
      onLastClick,
      onOlderClick,
    } = this.props;

    const { columns, names, onItemClick } = createChartData({
      last,
      older,
      timePeriod,
      onLastClick,
      onOlderClick,
    });

    const classes = classNames(
      'donut-chart-pf',
      'tasks-donut-chart',
      className,
      { 'tasks-donut-selected': shouleBeSelected(focusedOn) }
    );

    return (
      <Chart
        {...baseChartConfig}
        type="donut"
        className={classes}
        color={{ pattern: colorsPattern }}
        onChartCreate={chart => this.onChartCreate(chart)}
        data={{
          columns,
          names,
          onclick: ({ id }) => onItemClick(id),
          onmouseover: ({ id }) => this.setChartFocus(id),
          onmouseout: () => this.updateChartFocus(),
          type: 'donut',
        }}
        legend={{
          show: true,
          position: 'right',
          item: {
            onclick: id => onItemClick(id),
            onmouseover: id => this.setChartFocus(id),
            onmouseout: () => this.updateChartFocus(),
          },
        }}
      />
    );
  }
}

TasksDonutChart.propTypes = {
  last: PropTypes.number.isRequired,
  older: PropTypes.number.isRequired,
  className: PropTypes.string,
  timePeriod: PropTypes.string,
  colorsPattern: PropTypes.arrayOf(PropTypes.string),
  focusedOn: PropTypes.oneOf(TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS_ARRAY),
  onTotalClick: PropTypes.func,
  onLastClick: PropTypes.func,
  onOlderClick: PropTypes.func,
};

TasksDonutChart.defaultProps = {
  className: '',
  timePeriod: '24h',
  colorsPattern: COLLOR_PATTERN,
  focusedOn: TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.NORMAL,
  onTotalClick: () => null,
  onLastClick: () => null,
  onOlderClick: () => null,
};

export default TasksDonutChart;
