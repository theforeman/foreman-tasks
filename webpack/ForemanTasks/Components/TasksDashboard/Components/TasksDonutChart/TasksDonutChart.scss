.tasks-donut-chart {

  .c3-chart-arcs-title {
    cursor: pointer;

    &:hover {
      .donut-title-big-pf, .donut-title-small-pf {
        font-weight: bold;
      }
    }
  }

  &.tasks-donut-selected {
    .c3-chart-arcs-title {
      .donut-title-big-pf, .donut-title-small-pf {
        font-weight: bold;
      }
    }
  }
}
