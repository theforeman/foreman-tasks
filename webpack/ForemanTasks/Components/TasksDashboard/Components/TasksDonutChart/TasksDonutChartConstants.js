export const TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS = {
  NORMAL: 'normal', // normal-mode
  TOTAL: 'total', // total-mode
  LAST: 'last', // last X mode
  OLDER: 'older', // older then X mode
  NONE: 'none', // unfocus-mode: another card is selected
};

export const TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS_ARRAY = Object.values(
  TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS
);

export const COLLOR_PATTERN = ['#C315C7', '#0089C9'];
