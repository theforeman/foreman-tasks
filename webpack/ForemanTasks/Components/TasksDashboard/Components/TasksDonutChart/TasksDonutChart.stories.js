import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, number, text, select } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import { TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS_ARRAY } from './TasksDonutChartConstants';
import TasksDonutChart from './TasksDonutChart';

storiesOf('TasksDashboard', module)
  .addDecorator(withKnobs)
  .add('TasksDonutChart', () => (
    <TasksDonutChart
      last={number('last', 3)}
      older={number('older', 5)}
      time={text('time', '24h')}
      focusedOn={select(
        'focusedOn',
        TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS_ARRAY,
        TasksDonutChart.defaultProps.focusedOn
      )}
      onTotalClick={action('onTotalClick')}
      onLastClick={action('onLastClick')}
      onOlderClick={action('onOlderClick')}
    />
  ));
