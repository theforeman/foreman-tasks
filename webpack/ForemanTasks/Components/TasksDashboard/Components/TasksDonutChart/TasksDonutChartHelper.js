import { translate as __, sprintf } from 'foremanReact/common/I18n';
import { TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS } from './TasksDonutChartConstants';

const { patternfly } = window;

export const baseChartConfig = patternfly
  .c3ChartDefaults()
  .getDefaultDonutConfig();

export const shouleBeSelected = focusedOn =>
  focusedOn !== TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.NORMAL &&
  focusedOn !== TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.NONE;

export const createChartData = ({
  last,
  older,
  timePeriod,
  onLastClick,
  onOlderClick,
}) => {
  const lastKey = TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.LAST;
  const olderKey = TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS.OLDER;

  const data = {
    [lastKey]: {
      name: sprintf(__('%(last)s Last %(timePeriod)s'), { last, timePeriod }),
      value: last,
      onClick: onLastClick,
    },
    [olderKey]: {
      name: sprintf(__('%(older)s Older %(timePeriod)s'), {
        older,
        timePeriod,
      }),
      value: older,
      onClick: onOlderClick,
    },
  };

  const columns = Object.entries(data).map(([key, item]) => [key, item.value]);
  const names = {
    [lastKey]: data[lastKey].name,
    [olderKey]: data[olderKey].name,
  };
  const onItemClick = key => data[key].onClick();

  return { columns, names, onItemClick };
};

export const updateChartTitle = ({
  chartElement,
  value,
  onClick,
  onMouseOver,
  onMouseOut,
}) => {
  patternfly.pfSetDonutChartTitle(chartElement, value, __('Total'));

  window.d3
    .select(chartElement)
    .select('text.c3-chart-arcs-title')
    .on('click', onClick)
    .on('mouseover', onMouseOver)
    .on('mouseout', onMouseOut);
};
