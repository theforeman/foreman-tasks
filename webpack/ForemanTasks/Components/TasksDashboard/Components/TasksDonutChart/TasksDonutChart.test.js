import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import { TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS_ARRAY } from './TasksDonutChartConstants';
import TasksDonutChart from './TasksDonutChart';

jest.mock('./TasksDonutChartHelper', () => ({
  shouleBeSelected: focusedOn => focusedOn !== 'normal' && focusedOn !== 'none',
  baseChartConfig: () => ({ base: 'some-base-config' }),
  createChartData: jest.fn(() => ({
    columns: 'some-columns',
    names: 'some-names',
    onItemClick: jest.fn(),
  })),
  updateChartTitle: jest.fn(),
}));

const createRequiredProps = () => ({ last: 3, older: 5 });

const fixtures = {
  'render with minimal props': { ...createRequiredProps() },
  'render with props': {
    ...createRequiredProps(),
    className: 'some-class',
    time: 'time-period',
    colorsPattern: ['color1', 'color2'],
    onTotalClick: jest.fn(),
    onLastClick: jest.fn(),
    onOlderClick: jest.fn(),
  },
};

TASKS_DONUT_CHART_FOCUSED_ON_OPTIONS_ARRAY.forEach(mode => {
  fixtures[`render with focused-on ${mode}`] = {
    ...createRequiredProps(),
    focusedOn: mode,
  };
});

describe('TasksDonutChart', () =>
  testComponentSnapshotsWithFixtures(TasksDonutChart, fixtures));
