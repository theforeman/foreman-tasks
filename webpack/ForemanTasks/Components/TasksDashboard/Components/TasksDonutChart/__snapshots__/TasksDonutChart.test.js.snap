// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksDonutChart render with focused-on last 1`] = `
<C3Chart
  className="donut-chart-pf tasks-donut-chart tasks-donut-selected"
  color={
    Object {
      "pattern": Array [
        "#C315C7",
        "#0089C9",
      ],
    }
  }
  data={
    Object {
      "columns": "some-columns",
      "names": "some-names",
      "onclick": [Function],
      "onmouseout": [Function],
      "onmouseover": [Function],
      "type": "donut",
    }
  }
  legend={
    Object {
      "item": Object {
        "onclick": [Function],
        "onmouseout": [Function],
        "onmouseover": [Function],
      },
      "position": "right",
      "show": true,
    }
  }
  onChartCreate={[Function]}
  type="donut"
/>
`;

exports[`TasksDonutChart render with focused-on none 1`] = `
<C3Chart
  className="donut-chart-pf tasks-donut-chart"
  color={
    Object {
      "pattern": Array [
        "#C315C7",
        "#0089C9",
      ],
    }
  }
  data={
    Object {
      "columns": "some-columns",
      "names": "some-names",
      "onclick": [Function],
      "onmouseout": [Function],
      "onmouseover": [Function],
      "type": "donut",
    }
  }
  legend={
    Object {
      "item": Object {
        "onclick": [Function],
        "onmouseout": [Function],
        "onmouseover": [Function],
      },
      "position": "right",
      "show": true,
    }
  }
  onChartCreate={[Function]}
  type="donut"
/>
`;

exports[`TasksDonutChart render with focused-on normal 1`] = `
<C3Chart
  className="donut-chart-pf tasks-donut-chart"
  color={
    Object {
      "pattern": Array [
        "#C315C7",
        "#0089C9",
      ],
    }
  }
  data={
    Object {
      "columns": "some-columns",
      "names": "some-names",
      "onclick": [Function],
      "onmouseout": [Function],
      "onmouseover": [Function],
      "type": "donut",
    }
  }
  legend={
    Object {
      "item": Object {
        "onclick": [Function],
        "onmouseout": [Function],
        "onmouseover": [Function],
      },
      "position": "right",
      "show": true,
    }
  }
  onChartCreate={[Function]}
  type="donut"
/>
`;

exports[`TasksDonutChart render with focused-on older 1`] = `
<C3Chart
  className="donut-chart-pf tasks-donut-chart tasks-donut-selected"
  color={
    Object {
      "pattern": Array [
        "#C315C7",
        "#0089C9",
      ],
    }
  }
  data={
    Object {
      "columns": "some-columns",
      "names": "some-names",
      "onclick": [Function],
      "onmouseout": [Function],
      "onmouseover": [Function],
      "type": "donut",
    }
  }
  legend={
    Object {
      "item": Object {
        "onclick": [Function],
        "onmouseout": [Function],
        "onmouseover": [Function],
      },
      "position": "right",
      "show": true,
    }
  }
  onChartCreate={[Function]}
  type="donut"
/>
`;

exports[`TasksDonutChart render with focused-on total 1`] = `
<C3Chart
  className="donut-chart-pf tasks-donut-chart tasks-donut-selected"
  color={
    Object {
      "pattern": Array [
        "#C315C7",
        "#0089C9",
      ],
    }
  }
  data={
    Object {
      "columns": "some-columns",
      "names": "some-names",
      "onclick": [Function],
      "onmouseout": [Function],
      "onmouseover": [Function],
      "type": "donut",
    }
  }
  legend={
    Object {
      "item": Object {
        "onclick": [Function],
        "onmouseout": [Function],
        "onmouseover": [Function],
      },
      "position": "right",
      "show": true,
    }
  }
  onChartCreate={[Function]}
  type="donut"
/>
`;

exports[`TasksDonutChart render with minimal props 1`] = `
<C3Chart
  className="donut-chart-pf tasks-donut-chart"
  color={
    Object {
      "pattern": Array [
        "#C315C7",
        "#0089C9",
      ],
    }
  }
  data={
    Object {
      "columns": "some-columns",
      "names": "some-names",
      "onclick": [Function],
      "onmouseout": [Function],
      "onmouseover": [Function],
      "type": "donut",
    }
  }
  legend={
    Object {
      "item": Object {
        "onclick": [Function],
        "onmouseout": [Function],
        "onmouseover": [Function],
      },
      "position": "right",
      "show": true,
    }
  }
  onChartCreate={[Function]}
  type="donut"
/>
`;

exports[`TasksDonutChart render with props 1`] = `
<C3Chart
  className="donut-chart-pf tasks-donut-chart some-class"
  color={
    Object {
      "pattern": Array [
        "color1",
        "color2",
      ],
    }
  }
  data={
    Object {
      "columns": "some-columns",
      "names": "some-names",
      "onclick": [Function],
      "onmouseout": [Function],
      "onmouseover": [Function],
      "type": "donut",
    }
  }
  legend={
    Object {
      "item": Object {
        "onclick": [Function],
        "onmouseout": [Function],
        "onmouseover": [Function],
      },
      "position": "right",
      "show": true,
    }
  }
  onChartCreate={[Function]}
  type="donut"
/>
`;
